docs/tutorial/programming.md

```diff
@@ -1,10 +1,295 @@
-# Programming your Design
+# Programming Your Design
 
-# TODO:
-- Provide step-by-step programming
+Once the build is finished we are now ready to program our accelerator! The only thing we need to keep track of is the set registers that our accelerators are using. The major tasks in this section are to:
 
-# Make the C-Code
+1. Make a program in C.
+2. Generate a data header file using a data generation script.
+3. Modify `Makefiles` for the build.
+4. Building the program.
+5. Run the program.
 
-# Some Useful Functions
+We will make a simple program that does the following:
+
+1. Load data from the external memory to the TCDM memory.
+2. Configure the streamer registers and start the streamer.
+3. Configure the SNAX ALU and start the accelerator.
+4. Check if the data is correct.
+5. Print results.
+
+# Make the C-Code for SNAX ALU
+
+You can find all application programs in `./target/snitch_cluster/sw/apps/.`. Currently, there are several existing programs, but we'll look into the `snax-alu` directory. You should see the tree:
+
+```
+├── target/snitch_cluster/sw/apps
+│   ├── snax-alu
+│   |   ├── data
+|   |   |   ├── datagen.py
+|   |   |   └── Makefile
+|   |   ├── src
+|   |   |   └── snax-alu.c
+│   |   └── Makefile
+```
+
+Let's study the C program under the `./target/snitch_cluster/sw/apps/snax-alu/src/snax-alu.c`. First let's look at the included header files:
+
+```C
+#include "snrt.h"
+#include "data.h"
+```
+
+The `snrt.h` is a Snitch run-time header file for all pre-built functions made by the Snitch platform. You can find details in `./target/snitch_cluster/sw/runtime/rtl/src/snrt.h`. However, several of the definitions inside `snrt.h` are located in `./sw/snRuntime/src/.`. The `data.h` will contain the data that we need later. This will be automatically generated by the `Makefile`. 
+
+Inside the `main()` function we begin with declaring the pointers for the data addresses. Consider the code snippet below: 
+
+```C
+// Allocates space in TCDM
+uint64_t *local_a, *local_b, *local_o;
+
+local_a = (uint64_t *)snrt_l1_next();
+local_b = local_a + DATA_LEN;
+local_o = local_b + DATA_LEN;
+```
+We will use `a` and `b` for inputs and `o` for the output. In the code snippet below the `snrt_l1_next()` is a built-in Snitch run-time function that assigns the start of the base address of the TCDM. You can find the function definition in `./sw/snRuntime/src/alloc.h`. The `DATA_LEN` is a `uint64_t` data indicating the total number of elements we need to process in our program. You can find this in the generated `data.h` later.
+
+The first section is to run the Snitch with a DMA core to transfer data from an external memory (L2) to the local TCDM (L1) memory. See the code snippet below:
```

Comment:
```suggestion
The first section is to run the Snitch with a DMA core to transfer data from an external memory (L3) to the local TCDM (L1) memory. See the code snippet below:
```
This is referred to as L3 in various parts. L2 will be the host cores private/shared memory

Reply:
Ahh you're right okie!